File: src/frontend/js/solver.js
/**
 * solver.js
 * 
 * This object literal implements numerical solvers of the Runge-Kutta family.
 * Notes: the output of the solvers include 'v' key for velocity, but is unused at present
 */

import { Solution } from './solution.js'

// Helper functions
function arrAdd(u, v) {
    if (u.length === v.length) {
        let sum = [];
        for (let i = 0; i < u.length; i++) {
            sum.push(u[i] + v[i]);
        }
        return sum;
    } else {
        throw new Error("Mismatched array dimensions");
    }
}
function arrDiff(u, v) {
    if (u.length === v.length) {
        let diff = [];
        for (let i = 0; i < u.length; i++) {
            diff.push(u[i] - v[i]);
        }
        return diff;
    } else {
        throw new Error("Mismatched array dimensions");
    }
}
function arrMult(u, b) {
    let result = [];
    for (let i = 0; i < u.length; i++) {
        result.push(b*u[i]);
    }
    return result;
}
function arrDotMat(M, x) {
    let sum = [0, 0];
    for (let i = 0; i < M.length; i++) {
        sum = arrAdd(sum, arrMult(M[i], x[i]));
    }
    return sum;
}

export const Solver = {
   
    /**
     * Exact Analytical Solver for Damped Harmonic Oscillator
     * @param {number} t - Target evaluation time
     * @param {object} params - Input configuration { x0, v0, m, k, b }
     * @returns {Array<number>} - Returns state array [position, velocity]
     */
    calcState(t, params) {
        const { x0, v0, m, k, b, ts, t_end } = params;

        let discriminant = b*b - 4*m*k;
        
        if (discriminant > 0) {
            // Overdamped case
            let r1 = (-b + Math.sqrt(discriminant)) / (2*m);
            let r2 = (-b - Math.sqrt(discriminant)) / (2*m);
            let a1 = (r2*x0 - v0) / (r2 - r1);
            let a2 = x0 - a1;
            let x = a1*Math.exp(r1*t) + a2*Math.exp(r2*t);
            let v = a1*r1*Math.exp(r1*t) + a2*r2*Math.exp(r2*t);
            return [x, v]; 
        } else if (discriminant === 0) {
            // Critically damped case
            let r = -b / (2*m);
            let a1 = x0;
            let a2 = v0 - a1*r;
            let x = (a1 + a2*t) * Math.exp(r*t);
            let v = (a1 + a2*t) * r*Math.exp(r*t) + a2*Math.exp(r*t);
            return [x, v];
        } else {
            // Underdamped case
            let omega = Math.sqrt(4 * m * k - b * b) / (2 * m);
            let alpha = -b / (2 * m);
            let a1 = (v0 - alpha * x0) / omega;
            let a2 = x0;
            let x = Math.exp(alpha * t) * (a2 * Math.cos(omega * t) + a1 * Math.sin(omega * t));
            let v = Math.exp(alpha * t) * (a2 * (alpha * Math.cos(omega * t) - omega * Math.sin(omega * t)) + a1 * (alpha * Math.sin(omega * t) + omega * Math.cos(omega * t)));
            return [x, v];
        }
    },

    /**
     * Coupled State-Space Derivative Function f(t, x) for Damped Harmonic Oscillator
     * @param {number} t - Current time
     * @param {Array<number>} x - State array [position, velocity]
     * @param {object} params - Oscillator parameters
     * @returns {Array<number>} - Derivative array [dx_dt, dv_dt]
     */
    calcDerivatives(t, x, params) {
        // Destructure the 2-element state array and let x1 = x, x2 = v
        const [x1, x2] = x; 
        const { x0, v0, m, k, b, ts, t_end } = params;
        
        // State space formulation
        const dx1_dt = x2;
        const dx2_dt = (-k*x1 - b*x2) / m;

        return [dx1_dt, dx2_dt];
    },

    /**
     * Exact Analytical Solver
     * @param {object} params - Input configuration { x0, v0, m, k, b }
     * @returns {Array<object>} - Array of states [{ t, x, v }, ...]
     */
    solveExact(params) {
        const { x0, v0, m, k, b, ts, t_end } = params;
        const dt = Math.max(0.01, ts / 5); // Ensure a minimum dt for smoothness
        const steps = Math.floor(t_end/dt);
        
        let t = 0;
        let x = [x0, v0];
        let y_hat = 0;

        
        const results = [];
        results.push({ t: 0, x: x[0], v: x[1], e: y_hat });

        // Loop over the total simulation steps
        for (let i = 1; i <= steps; i++) {
            t += dt;
            x = Solver.calcState(t, { x0, v0, m, k, b });
            results.push({ t: t, x: x[0], v: x[1], e: y_hat });
        }

        return new Solution(results);
    },

    /**
     * Forward Euler Method
     * @param {object} params - Configuration { x0, v0, ts, t_end, m, k, b }
     * @returns {Array<object>} - Array of states [{ t, x, v }, ...]
     */
    solveEuler(params) {
        const { x0, v0, m, k, b, ts, t_end } = params;
        const steps = Math.floor(t_end/ts);
        
        let t = 0;
        let x = [x0, v0];
        let y_hat = 0;
        
        const results = [];
        results.push({ t: 0, x: x[0], v: x[1], e: y_hat });

        // Loop over the total simulation steps
        for (let i = 1; i <= steps; i++) {
            const dx_dt = Solver.calcDerivatives(t, x, { m, k, b });
            x[0] += ts * dx_dt[0];
            x[1] += ts * dx_dt[1];

            t += ts;
            y_hat = x[0] - Solver.calcState(t, { x0, v0, m, k, b })[0];
            results.push({ t: t, x: x[0], v: x[1], e: y_hat });
        }

        return new Solution(results);
    },
    
    /**
     * Runge-Kutta 2nd Order Method (RK2) 
     * @param {object} params  - Configuration { x0, v0, ts, t_end, m, k, b }
     * @returns {Array<object>} - Array of states [{ t, x, v }, ...]
     */
    solveRK2(params) {
        const { x0, v0, m, k, b, ts, t_end } = params;
        const steps = Math.floor(t_end/ts);
        
        let t = 0;
        let x = [x0, v0];
        let y_hat = 0;
        
        const results = [];
        results.push({ t: 0, x: x[0], v: x[1], e: y_hat });

        // Loop over the total simulation steps
        for (let i = 1; i <= steps; i++) {
            const dx_dt = Solver.calcDerivatives(t, x, { m, k, b });
            let k1 = dx_dt;
            let k2 = Solver.calcDerivatives(t + ts/2, [x[0] + ts/2 * k1[0], x[1] + ts/2 * k1[1]], { m, k, b });
            x[0] += ts * k2[0];
            x[1] += ts * k2[1];

            t += ts;
            y_hat = x[0] - Solver.calcState(t, { x0, v0, m, k, b })[0];
            results.push({ t: t, x: x[0], v: x[1], e: y_hat });
        }

        return new Solution(results);
    },
    
    /**
     * Runge-Kutta 4th Order Method (RK4) 
     * @param {object} params - Configuration { x0, v0, ts, t_end, m, k, b }
     * @returns {Array<object>} - Array of states [{ t, x, v }, ...]
     */
    solveRK4(params) {
        const { x0, v0, m, k, b, ts, t_end } = params;
        const steps = Math.floor(t_end/ts);
        
        let t = 0;
        let x = [x0, v0];
        let y_hat = 0;
        
        const results = [];
        results.push({ t: 0, x: x[0], v: x[1], e: y_hat });

        // Loop over the total simulation steps
        for (let i = 1; i <= steps; i++) {
            const dx_dt = Solver.calcDerivatives(t, x, { m, k, b });
            let k1 = dx_dt;
            let k2 = Solver.calcDerivatives(t + ts/2, [x[0] + ts/2 * k1[0], x[1] + ts/2 * k1[1]], { m, k, b });
            let k3 = Solver.calcDerivatives(t + ts/2, [x[0] + ts/2 * k2[0], x[1] + ts/2 * k2[1]], { m, k, b });
            let k4 = Solver.calcDerivatives(t + ts, [x[0] + ts * k3[0], x[1] + ts * k3[1]], { m, k, b });
            x[0] += ts/6 * (k1[0] + 2*k2[0] + 2*k3[0] + k4[0]);
            x[1] += ts/6 * (k1[1] + 2*k2[1] + 2*k3[1] + k4[1]);

            t += ts;
            y_hat = x[0] - Solver.calcState(t, { x0, v0, m, k, b })[0];
            results.push({ t: t, x: x[0], v: x[1], e: y_hat });
        }

        return new Solution(results);
    },

    solveDP45(params) {

        const { x0, v0, m, k, b, ts, t_end } = params;

        const sf = 0.9; // Safety factor
        const tol = 1e-4; // Tolerance for error control

        let t = 0;
        let x = [x0, v0];
        let y_hat = 0;
        let dt = ts;

        const results = [];
        results.push({ t: 0, x: x[0], v: x[1], e: y_hat });

        const tableu = {
            a: [
                [],
                [1/5],
                [3/40, 9/40],
                [44/45, -56/15, 32/9],
                [19372/6561, -25360/2187, 64448/6561, -212/729],
                [9017/3168, -355/33, 46732/5247, 49/176, -5103/18656],
                [35/384, 0, 500/1113, 125/192, -2187/6784, 11/84]
            ],
            b: [35/384, 0, 500/1113, 125/192, -2187/6784, 11/84, 0],
            b_hat: [5179/57600, 0, 7571/16695, 393/640, -92097/339200, 187/2100, 1/40],
            c: [0, 1/5, 3/10, 4/5, 8/9, 1, 1]
        };

        while (t < t_end) {
            dt = Math.min(dt, t_end - t);

            let ks = [];

            for (let i = 0; i < tableu.c.length; i++) {
                let Deltax = [0, 0];

                if (i === 0) {
                    ks.push(Solver.calcDerivatives(t, x, params));
                } else {
                    for (let j = 0; j < i; j++) {
                        Deltax = arrAdd(arrMult(ks[j], tableu.a[i][j]*dt), Deltax);
                    }
                    ks.push(Solver.calcDerivatives(t + tableu.c[i]*dt, arrAdd(x, Deltax), params));
                }
            }

            let dx = arrDotMat(ks, arrMult(tableu.b, dt));
            let dx_hat = arrDotMat(ks, arrMult(tableu.b_hat, dt));

            let error = Math.hypot(...arrDiff(dx, dx_hat));

            if (error <= tol) {
                x = arrAdd(x, dx);
                t += dt;
                y_hat = x[0] - Solver.calcState(t, { x0, v0, m, k, b })[0];
                results.push({ t: t, x: x[0], v: x[1], e: y_hat });
            }
            dt *= sf*Math.pow(tol/error, 1/5);
        }

        return new Solution(results);
    }

}

File: src/backend/solver.py
import numpy as np
from numba import njit
from .physics import calc_exact_state, calc_derivatives

@njit
def solve_exact(x0, v0, m, k, b, ts, t_end):
    """
    Exact Analytical Solution sampled at fine grid for smooth display.
    """
    dt = max(0.01, ts / 5.0)
    steps = int(np.floor(t_end / dt))
    
    results = np.zeros((steps + 1, 4))
    results[0, 0] = 0.0
    results[0, 1] = x0
    results[0, 2] = v0
    results[0, 3] = 0.0
    
    t = 0.0
    for i in range(1, steps + 1):
        t += dt
        state = calc_exact_state(t, x0, v0, m, k, b)
        results[i, 0] = t
        results[i, 1] = state[0]
        results[i, 2] = state[1]
        results[i, 3] = 0.0
        
    return results

@njit
def solve_euler(x0, v0, m, k, b, ts, t_end):
    """
    Forward Euler integration.
    """
    steps = int(np.floor(t_end / ts))
    results = np.zeros((steps + 1, 4))
    
    results[0, 0] = 0.0
    results[0, 1] = x0
    results[0, 2] = v0
    results[0, 3] = 0.0
    
    t = 0.0
    state = np.array([x0, v0], dtype=np.float64)
    
    for i in range(1, steps + 1):
        dydt = calc_derivatives(t, state, m, k, b)
        state[0] += ts * dydt[0]
        state[1] += ts * dydt[1]
        t += ts
        
        exact = calc_exact_state(t, x0, v0, m, k, b)
        error = state[0] - exact[0]
        
        results[i, 0] = t
        results[i, 1] = state[0]
        results[i, 2] = state[1]
        results[i, 3] = error
        
    return results

@njit
def solve_rk2(x0, v0, m, k, b, ts, t_end):
    """
    Second-Order Runge-Kutta (Midpoint Method).
    """
    steps = int(np.floor(t_end / ts))
    results = np.zeros((steps + 1, 4))
    
    results[0, 0] = 0.0
    results[0, 1] = x0
    results[0, 2] = v0
    results[0, 3] = 0.0
    
    t = 0.0
    state = np.array([x0, v0], dtype=np.float64)
    
    for i in range(1, steps + 1):
        k1 = calc_derivatives(t, state, m, k, b)
        
        state_half = np.array([state[0] + 0.5 * ts * k1[0], state[1] + 0.5 * ts * k1[1]], dtype=np.float64)
        k2 = calc_derivatives(t + 0.5 * ts, state_half, m, k, b)
        
        state[0] += ts * k2[0]
        state[1] += ts * k2[1]
        t += ts
        
        exact = calc_exact_state(t, x0, v0, m, k, b)
        error = state[0] - exact[0]
        
        results[i, 0] = t
        results[i, 1] = state[0]
        results[i, 2] = state[1]
        results[i, 3] = error
        
    return results

@njit
def solve_rk4(x0, v0, m, k, b, ts, t_end):
    """
    Fourth-Order Runge-Kutta (Classical RK4).
    """
    steps = int(np.floor(t_end / ts))
    results = np.zeros((steps + 1, 4))
    
    results[0, 0] = 0.0
    results[0, 1] = x0
    results[0, 2] = v0
    results[0, 3] = 0.0
    
    t = 0.0
    state = np.array([x0, v0], dtype=np.float64)
    
    for i in range(1, steps + 1):
        k1 = calc_derivatives(t, state, m, k, b)
        
        state_k2 = np.array([state[0] + 0.5 * ts * k1[0], state[1] + 0.5 * ts * k1[1]], dtype=np.float64)
        k2 = calc_derivatives(t + 0.5 * ts, state_k2, m, k, b)
        
        state_k3 = np.array([state[0] + 0.5 * ts * k2[0], state[1] + 0.5 * ts * k2[1]], dtype=np.float64)
        k3 = calc_derivatives(t + 0.5 * ts, state_k3, m, k, b)
        
        state_k4 = np.array([state[0] + ts * k3[0], state[1] + ts * k3[1]], dtype=np.float64)
        k4 = calc_derivatives(t + ts, state_k4, m, k, b)
        
        state[0] += (ts / 6.0) * (k1[0] + 2.0 * k2[0] + 2.0 * k3[0] + k4[0])
        state[1] += (ts / 6.0) * (k1[1] + 2.0 * k2[1] + 2.0 * k3[1] + k4[1])
        t += ts
        
        exact = calc_exact_state(t, x0, v0, m, k, b)
        error = state[0] - exact[0]
        
        results[i, 0] = t
        results[i, 1] = state[0]
        results[i, 2] = state[1]
        results[i, 3] = error
        
    return results

@njit
def solve_dp45(x0, v0, m, k, b, ts, t_end):
    """
    Runge-Kutta-Fehlberg 4/5 (Dormand-Prince adaptive time step method).
    """
    sf = 0.9
    tol = 1e-4
    
    t = 0.0
    state = np.array([x0, v0], dtype=np.float64)
    dt = ts
    
    # Butcher tableau values
    c = np.array([0.0, 1/5, 3/10, 4/5, 8/9, 1.0, 1.0])
    
    a = np.zeros((7, 7))
    a[1, 0] = 1/5
    a[2, 0] = 3/40
    a[2, 1] = 9/40
    a[3, 0] = 44/45
    a[3, 1] = -56/15
    a[3, 2] = 32/9
    a[4, 0] = 19372/6561
    a[4, 1] = -25360/2187
    a[4, 2] = 64448/6561
    a[4, 3] = -212/729
    a[5, 0] = 9017/3168
    a[5, 1] = -355/33
    a[5, 2] = 46732/5247
    a[5, 3] = 49/176
    a[5, 4] = -5103/18656
    a[6, 0] = 35/384
    a[6, 1] = 0.0
    a[6, 2] = 500/1113
    a[6, 3] = 125/192
    a[6, 4] = -2187/6784
    a[6, 5] = 11/84
    
    b_sol = np.array([35/384, 0.0, 500/1113, 125/192, -2187/6784, 11/84, 0.0])
    b_hat = np.array([5179/57600, 0.0, 7571/16695, 393/640, -92097/339200, 187/2100, 1/40])
    
    # Allocate a large array for adaptive output
    max_steps = 20000
    results = np.zeros((max_steps, 4))
    
    results[0, 0] = 0.0
    results[0, 1] = x0
    results[0, 2] = v0
    results[0, 3] = 0.0
    
    step_idx = 1
    
    while t < t_end and step_idx < max_steps:
        dt = min(dt, t_end - t)
        
        # Calculate stages
        ks = np.zeros((7, 2))
        for i in range(7):
            delta_x = np.zeros(2)
            for j in range(i):
                delta_x += ks[j] * (a[i, j] * dt)
            
            ks[i] = calc_derivatives(t + c[i] * dt, state + delta_x, m, k, b)
            
        # Calculate the 4th and 5th order step updates
        dx = np.zeros(2)
        dx_hat = np.zeros(2)
        for i in range(7):
            dx += ks[i] * (b_sol[i] * dt)
            dx_hat += ks[i] * (b_hat[i] * dt)
            
        # Difference gives the error estimate
        diff = dx - dx_hat
        error = np.sqrt(diff[0]**2 + diff[1]**2)
        
        # Accept step if error is within tolerance
        if error <= tol:
            state += dx
            t += dt
            
            exact = calc_exact_state(t, x0, v0, m, k, b)
            y_hat = state[0] - exact[0]
            
            results[step_idx, 0] = t
            results[step_idx, 1] = state[0]
            results[step_idx, 2] = state[1]
            results[step_idx, 3] = y_hat
            step_idx += 1
            
        # Update time step
        factor = tol / max(error, 1e-15)
        dt *= sf * (factor ** 0.2)
        
        # Avoid extremely small time steps
        if dt < 1e-6:
            dt = 1e-6
            
    return results[:step_idx]
